alternator/rjson.cc

```diff
@@ -44,7 +68,7 @@ rjson::value parse(const std::string& str) {
 
 rjson::value parse_raw(const char* c_str, size_t size) {
     rjson::document d;
-    d.Parse(c_str, size);
+    d.Parse<rapidjson::kParseIterativeFlag>(c_str, size);
```

Comment:
What's this? Something to get rapidjson to notice the macro?

Is a document with 1 million items allowed? Although I agree that parsing should be preemptible.

Reply:
It's actually a change that could be merged separately anyway - the default implementation is callstack-based, so parsing large items (in terms of depth) without this flag may cause problems. I think that alternator limits the requests to <400KiB, but that could still store several thousands of small items, each being a parent of another.

Reply:
From rapidjson (http://rapidjson.org/md_doc_features.html):
```
Recursive (default) and iterative parser
    Recursive parser is faster but prone to stack overflow in extreme cases.
    Iterative parser use custom stack to keep parsing state.

```
I also wanted to create a test in the alternator-test suite for deeply nested objects, but boto3's Python very quickly complains that I reached a max level of recursion (`sys.getrecursionlimit()`, 3000 frames in my configuration) - and a dictionary of depth 150(sic!) is already able to hit that. It's not hard to create a deeply nested JSON in a long enough string, but I also think that this limit in boto3 may be a strong indicator that users don't often use such deeply nested structures, except for attacks.

So, perhaps we should just guard against too large objects and use the default recursive implementation, given that the authors claim it's faster? Naturally, a benchmark could also help.

Reply:
I opened https://github.com/scylladb/scylla/issues/5842. I think this should be discussed and patched separately.